File: 01-Numerical_Errors/3-complex_numbers.cpp
/* Complex numbers and functions in c++ */
#include<iostream>
#include<complex>
#include<iomanip>
#include<cmath>

using namespace std::complex_literals;

int main(int arcg, char **argv)
{
  std::cout << std::fixed << std::setprecision(2);

  //Decare complex numbers
  std::complex<double> z , e, ln, root;
  double arctan;

  //Declare norm and phase
  double r;
  double phi;

  std::cout << "phi" << "\t" << "x" << "\t" << "y" << "\t"
            << "    e^z    " << "\t" << "   ln(z)   " << "\t"
            << "   sqrt(z) " << "\t" << " atan(y/x)   " << "\n";

  std::cout << "---------------------------------------------------------------------------------" << '\n';
  //Loop over phases
  for(phi = -4*M_PI; phi <= 4*M_PI; phi+= M_PI_4 )
  {
  //Assign a value to it
  r = 2;
  z = r*exp(1i * phi);

  //Compute real and imaginary parts
  double x = real(z);
  double y = imag(z);

  //Compute some complex functions
  e = exp(z);
  ln = log(z);
  root = sqrt(z);
  arctan = std::atan(y/x);

  std::cout << phi << "\t" << x << "\t" << y << "\t" << e << "\t" << ln << "\t"<< root << "\t" << arctan << "\n";

  }

  return 0;
}

File: 01-Numerical_Errors/3-complex_numbers.py
"""
Description: Complex numbers and functions in Python
"""
import numpy as np
import cmath as cm
import matplotlib.pyplot as plt


print("phi", "\t", " x ", " y ", "  e  ", "  ln  ", "  sqrt  ", "  atan  ")
print("--------------------------------------------------------")

for phi in np.arange(-4*np.pi, 4*np.pi, 0.25*np.pi):
    #Compute phase and norm
    r = 1

    #Declare complex number
    z = r*cm.exp(phi*1j)

    #Compute imaginary and real parts
    x = z.real
    y = z.imag

    #Compute functions
    e = cm.exp(z)
    ln = cm.log(z)
    root = cm.sqrt(z)
    atan = np.arctan(y/x)

    print(round(phi,2), round(x,2), round(y,2), e, ln, root, round(atan,2))


#Now, lets define vectors for making plots of the phase
InputPhases = np.arange(-4*np.pi, 4*np.pi, 0.25*np.pi)


AtanPhases = np.zeros(32)
for i in np.arange(32):
    r = 1
    z = r*cm.exp(InputPhases[i]*1j)
    x = z.real
    y = z.imag
    AtanPhases[i] = np.arctan(y/x)
    #print(InputPhases[i], AtanPhases[i])    

plt.title('Input phases vs atan')
plt.xlabel('input $\phi$')
plt.ylabel('$atan(y/x)$')

plt.plot(InputPhases,AtanPhases, 'y')    
plt.savefig('phases_py.pdf')    
    
    
"""
//////////////////////////////////////////////////////////////////////////
Created on Fri Mar  1 13:32:11 2021                                    //
                                                                      //
@author: Nicolás Gómez                                               //
//////////////////////////////////////////////////////////////////////
"""

    
